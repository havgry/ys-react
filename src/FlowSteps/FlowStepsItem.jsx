import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

const FlowStepsItem = (props) => {
  const {
    children,
    active,
    className,
  } = props

  const wrapperProps = {
    className: classNames(
      'ys-flow-steps__item',
      active && 'ys-flow-steps__item--active',
      className,
    ),
  }

  return (
    <li {...wrapperProps}>
      <span className="ys-flow-steps__label">
        {children}
      </span>
    </li>
  )
}

FlowStepsItem.propTypes = {
  children: PropTypes.string.isRequired,
  active: PropTypes.bool,
  className: PropTypes.string,
}

FlowStepsItem.defaultProps = {
  active: false,
  className: '',
}

export default FlowStepsItem
