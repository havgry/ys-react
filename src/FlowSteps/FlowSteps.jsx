import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import FlowStepsItem from './FlowStepsItem'

import componentPropType from '../../utils/componentPropType'

import '../base.css'
import '@youseedk/dna/css/elements/ys-flow-steps.css'

const FlowSteps = (props) => {
  const {
    children,
    className,
  } = props

  const wrapperProps = {
    className: classNames(
      'ys-flow-steps',
      className,
    ),
  }

  const totalSteps = children.length
  let activeStep = 0

  for (null; activeStep < totalSteps; activeStep += 1) {
    if (children[activeStep].props.active === true) {
      break
    }
  }

  const label = `${activeStep} ud af ${totalSteps} trin gennemfÃ¸rt`

  return (
    <div
      role="progressbar"
      aria-valuenow={activeStep}
      aria-valuemin={0}
      aria-valuemax={totalSteps}
      aria-valuetext={label}
      {...wrapperProps}
    >
      <ol className="ys-flow-steps__list" aria-hidden="true">
        {children}
      </ol>
    </div>
  )
}

FlowSteps.propTypes = {
  children: componentPropType(FlowStepsItem, true).isRequired,
  className: PropTypes.string,
}

FlowSteps.defaultProps = {
  className: '',
}

FlowSteps.Item = FlowStepsItem

export default FlowSteps
