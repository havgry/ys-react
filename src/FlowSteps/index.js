import FlowSteps from './FlowSteps'

export default FlowSteps
