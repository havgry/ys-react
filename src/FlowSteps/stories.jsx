import React from 'react'
import { storiesOf } from '@storybook/react'

import { text } from '@storybook/addon-knobs'

import FlowSteps from './FlowSteps'

storiesOf('FlowSteps', module)
  .addParameters({ jest: ['FlowSteps'] })
  .add('0% complete', () => (
    <FlowSteps
      className={text('className', '', 'Misc')}
    >
      <FlowSteps.Item active>Abonnement</FlowSteps.Item>
      <FlowSteps.Item>Adresse</FlowSteps.Item>
      <FlowSteps.Item>Forsendelse</FlowSteps.Item>
      <FlowSteps.Item>Bekræftelse</FlowSteps.Item>
      <FlowSteps.Item>Betaling</FlowSteps.Item>
    </FlowSteps>
  ))
  .add('40% complete', () => (
    <FlowSteps
      className={text('className', '', 'Misc')}
    >
      <FlowSteps.Item>Abonnement</FlowSteps.Item>
      <FlowSteps.Item>Adresse</FlowSteps.Item>
      <FlowSteps.Item active>Forsendelse</FlowSteps.Item>
      <FlowSteps.Item>Bekræftelse</FlowSteps.Item>
      <FlowSteps.Item>Betaling</FlowSteps.Item>
    </FlowSteps>
  ))
  .add('100% complete', () => (
    <FlowSteps
      className={text('className', '', 'Misc')}
    >
      <FlowSteps.Item>Abonnement</FlowSteps.Item>
      <FlowSteps.Item>Adresse</FlowSteps.Item>
      <FlowSteps.Item>Forsendelse</FlowSteps.Item>
      <FlowSteps.Item>Bekræftelse</FlowSteps.Item>
      <FlowSteps.Item>Betaling</FlowSteps.Item>
    </FlowSteps>
  ))
