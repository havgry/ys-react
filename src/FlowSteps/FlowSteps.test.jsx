import React from 'react'
import { shallow, mount } from 'enzyme'

import FlowSteps from './FlowSteps'

const customClassName = 'heron'

describe('<FlowSteps>', () => {
  describe('Props', () => {
    it('Doesn\'t throw propType warning if children are multiple `FlowSteps.Item` elements', () => {
      const component = () => shallow(
        <FlowSteps>
          <FlowSteps.Item>Step 1</FlowSteps.Item>
          <FlowSteps.Item>Step 2</FlowSteps.Item>
          <FlowSteps.Item>Step 3</FlowSteps.Item>
        </FlowSteps>,
      )

      expect(component).not.toThrow()
    })

    it('Throws propType warning if single child element is passed', () => {
      const component = () => shallow(
        <FlowSteps>
          <FlowSteps.Item>Step 1</FlowSteps.Item>
        </FlowSteps>,
      )
      expect(component).toThrow()
    })

    it('Throws propType warning if no `child` is passed', () => {
      const component = () => shallow(
        <FlowSteps />,
      )
      expect(component).toThrow()
    })

    it('Has correct defaultProps', () => {
      const component = mount(
        <FlowSteps>
          <FlowSteps.Item>Step 1</FlowSteps.Item>
        </FlowSteps>,
      )

      const {
        className,
      } = component.props()

      expect(className).toBe('')
    })
  })

  describe('Rendering', () => {
    it('ARIA attributes calculated and set correctly', () => {
      let component = shallow(
        <FlowSteps>
          <FlowSteps.Item active>Step 1</FlowSteps.Item>
          <FlowSteps.Item>Step 2</FlowSteps.Item>
          <FlowSteps.Item>Step 3</FlowSteps.Item>
        </FlowSteps>,
      )

      expect(component.prop('aria-valuenow')).toBe(0)
      expect(component.prop('aria-valuemin')).toBe(0)
      expect(component.prop('aria-valuemax')).toBe(3)

      component = shallow(
        <FlowSteps>
          <FlowSteps.Item>Step 1</FlowSteps.Item>
          <FlowSteps.Item active>Step 2</FlowSteps.Item>
          <FlowSteps.Item>Step 3</FlowSteps.Item>
        </FlowSteps>,
      )

      expect(component.prop('aria-valuenow')).toBe(1)

      component = shallow(
        <FlowSteps>
          <FlowSteps.Item>Step 1</FlowSteps.Item>
          <FlowSteps.Item>Step 2</FlowSteps.Item>
          <FlowSteps.Item active>Step 3</FlowSteps.Item>
        </FlowSteps>,
      )

      expect(component.prop('aria-valuenow')).toBe(2)

      component = shallow(
        <FlowSteps>
          <FlowSteps.Item>Step 1</FlowSteps.Item>
          <FlowSteps.Item>Step 2</FlowSteps.Item>
          <FlowSteps.Item>Step 3</FlowSteps.Item>
        </FlowSteps>,
      )

      expect(component.prop('aria-valuenow')).toBe(3)
    })

    it('Passes other generic props to container element', () => {
      const component = shallow(
        <FlowSteps tabIndex="0" id="test">
          <FlowSteps.Item>Step 1</FlowSteps.Item>
          <FlowSteps.Item>Step 2</FlowSteps.Item>
        </FlowSteps>,
      )

      expect(component.props().id).toBe('test')
      expect(component.props().tabIndex).toBe('0')
    })
  })

  describe('Snapshots', () => {
    it('With custom `className`', () => {
      const component = shallow(
        <FlowSteps className={customClassName}>
          <FlowSteps.Item>Step 1</FlowSteps.Item>
          <FlowSteps.Item>Step 2</FlowSteps.Item>
          <FlowSteps.Item>Step 3</FlowSteps.Item>
        </FlowSteps>,
      )

      expect(component).toMatchSnapshot()
    })

    it('0% progress', () => {
      const component = shallow(
        <FlowSteps>
          <FlowSteps.Item active>Step 1</FlowSteps.Item>
          <FlowSteps.Item>Step 2</FlowSteps.Item>
          <FlowSteps.Item>Step 3</FlowSteps.Item>
          <FlowSteps.Item>Step 4</FlowSteps.Item>
        </FlowSteps>,
      )

      expect(component).toMatchSnapshot()
    })

    it('25% progress', () => {
      const component = shallow(
        <FlowSteps>
          <FlowSteps.Item>Step 1</FlowSteps.Item>
          <FlowSteps.Item active>Step 2</FlowSteps.Item>
          <FlowSteps.Item>Step 3</FlowSteps.Item>
          <FlowSteps.Item>Step 4</FlowSteps.Item>
        </FlowSteps>,
      )

      expect(component).toMatchSnapshot()
    })

    it('100% progress', () => {
      const component = shallow(
        <FlowSteps>
          <FlowSteps.Item>Step 1</FlowSteps.Item>
          <FlowSteps.Item>Step 2</FlowSteps.Item>
          <FlowSteps.Item>Step 3</FlowSteps.Item>
          <FlowSteps.Item>Step 4</FlowSteps.Item>
        </FlowSteps>,
      )

      expect(component).toMatchSnapshot()
    })
  })
})
