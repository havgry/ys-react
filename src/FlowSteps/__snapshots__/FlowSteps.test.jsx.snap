// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlowSteps> Snapshots 0% progress 1`] = `
<div
  aria-valuemax={4}
  aria-valuemin={0}
  aria-valuenow={0}
  aria-valuetext="0 ud af 4 trin gennemført"
  className="ys-flow-steps"
  role="progressbar"
>
  <ol
    aria-hidden="true"
    className="ys-flow-steps__list"
  >
    <FlowStepsItem
      active={true}
      className=""
    >
      Step 1
    </FlowStepsItem>
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 2
    </FlowStepsItem>
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 3
    </FlowStepsItem>
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 4
    </FlowStepsItem>
  </ol>
</div>
`;

exports[`<FlowSteps> Snapshots 25% progress 1`] = `
<div
  aria-valuemax={4}
  aria-valuemin={0}
  aria-valuenow={1}
  aria-valuetext="1 ud af 4 trin gennemført"
  className="ys-flow-steps"
  role="progressbar"
>
  <ol
    aria-hidden="true"
    className="ys-flow-steps__list"
  >
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 1
    </FlowStepsItem>
    <FlowStepsItem
      active={true}
      className=""
    >
      Step 2
    </FlowStepsItem>
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 3
    </FlowStepsItem>
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 4
    </FlowStepsItem>
  </ol>
</div>
`;

exports[`<FlowSteps> Snapshots 100% progress 1`] = `
<div
  aria-valuemax={4}
  aria-valuemin={0}
  aria-valuenow={4}
  aria-valuetext="4 ud af 4 trin gennemført"
  className="ys-flow-steps"
  role="progressbar"
>
  <ol
    aria-hidden="true"
    className="ys-flow-steps__list"
  >
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 1
    </FlowStepsItem>
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 2
    </FlowStepsItem>
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 3
    </FlowStepsItem>
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 4
    </FlowStepsItem>
  </ol>
</div>
`;

exports[`<FlowSteps> Snapshots With custom \`className\` 1`] = `
<div
  aria-valuemax={3}
  aria-valuemin={0}
  aria-valuenow={3}
  aria-valuetext="3 ud af 3 trin gennemført"
  className="ys-flow-steps heron"
  role="progressbar"
>
  <ol
    aria-hidden="true"
    className="ys-flow-steps__list"
  >
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 1
    </FlowStepsItem>
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 2
    </FlowStepsItem>
    <FlowStepsItem
      active={false}
      className=""
    >
      Step 3
    </FlowStepsItem>
  </ol>
</div>
`;
